bbl_screen-patch/patches/printerui/qml/x1plus/GcodeGenerator.js

```diff
@@ -330,73 +254,299 @@ function G4(sec = 90) /* pause */
+        G1({e: 60, accel: 100}), // fill the nozzle and flush a very small amount
+        M106.aux(255), // fans on
+        M106.part(255),
+        M109(flush_temp), // lower temp
+        G1({e: 10, accel: 100}), // flush a small amount at a lower temp
+        M104(second_flush_temp) // lower temp
+    ].join(''),
+
+    pull: (pull_temp) => [
+        M109(pull_temp),
+        G1({e: -100, accel: 1200}) // extrude while the user pulls
+    ].join(''),
+
+    exit: () => [
+        M104(0),
+        M1002.gcode_claim_action(0),
+        M106.aux(0),
+        M106.part(0),
+        M84() // disable steppers
+    ].join('')
+};
+
+const Preheat = {
+    home: () => [
+        G28.z_low_precision(),
+        G91(), // relative
+        G0({z: 5, accel: 1200}) // move toolhead close to bed
+    ].join(''),
+
+    on: (temp) => [
+        M140(temp), // heater on
+        M106.aux(255), // fans on
+        M106.part(255)
+    ].join(''),
+
+    off: () => [
+        M140(0), // heater off
+        M106.aux(0), // fans off
+        M106.part(0),
+        G90() // absolute
+    ].join('')
+};
+
+
+const Tramming = {
+    exit: () => [
+        M1002.gcode_claim_action(254),
+        G1({x: 128, y: 128, z: 1}),
+        M400.M400(),
+        M1002.gcode_claim_action(0),
+    ].join(''),
+
+    prepare: () => [
+        M1002.gcode_claim_action(254),
+        M17(1.2, 1.2, 0.75),
+        G90(),
+        M83(),
+        G28.xyz(),
+        G1({x: 128, y: 128, z: 1}),
+        G292(0),
+        M1002.gcode_claim_action(1),
+    ].join(''),
+
+    rear_center: () => [
+        M1002.gcode_claim_action(254),
+        G1({x: 134.8, y: 242.8, z: 0.4, accel: 3600}),
+        M400.M400(),
+        M1002.gcode_claim_action(1),
+    ].join(''),
+
+    front_left: () => [
+        M1002.gcode_claim_action(254),
+        G1({x: 33.2, y: 13.2, z: 0.4, accel: 3600}),
+        M400.M400(),
+        M1002.gcode_claim_action(1),
+    ].join(''),
+
+    front_right: () => [
+        M1002.gcode_claim_action(254),
+        G1({x: 222.8, y: 13.2, z: 0.4, accel: 3600}),
+        M400.M400(),
+        M1002.gcode_claim_action(1),
+    ].join('')
+};
+
+/**  Speed ramping logic
+ * Input = current layer and speed ramp parameters, output = speed
+  */
+function getRampSpeed (currentLayer, startingLayer, targetLayer, startingSpeed, targetSpeed, delta) {
+    
+    if (startingSpeed > targetSpeed) {
+        delta = -Math.abs(delta);
+    }
+    if (currentLayer < startingLayer) {
+        return startingSpeed; 
+    } else if (currentLayer > targetLayer) {
+        return targetSpeed; 
+    } else {
+        const rampProgress = currentLayer - startingLayer;
+        let speedChange = rampProgress * delta;  
+        let newSpeed = startingSpeed + speedChange;
+        if (startingSpeed < targetSpeed) {
+            return Math.min(newSpeed, targetSpeed);
+        } else {
+            return Math.max(newSpeed, targetSpeed);
+        }
+    }    
+}
```

Comment:
This looks like a straightforward linear interpolator.  Should it be in GcodeGenerator or in whatever timer is actually doing the LERP?

Reply:
> This looks like a straightforward linear interpolator. Should it be in GcodeGenerator or in whatever timer is actually doing the LERP?

There's no timer used for the speed ramp feature. It uses an onLayerNumChanged signal in QML. Anyway I think we could find a better place for it than here, since it's not generating gcode

Reply:
Yeah, whatever the source is, I think the LERP belongs there